### PR TITLE
UCT/IB: Fix explicit ODP initialization

### PR DESCRIPTION
**Pull request by ivankochin**
## What
Fix the bug that causes absence of ODP registration even if `UCT_MD_MEM_FLAG_NONBLOCK` was passed to `uct_md_mem_reg`

## How ?
The current `uct_ib_mlx5_devx_md_open` flow that causes the issue:
1. `uct_ib_mlx5_devx_check_odp()` set `md->super.config.odp.max_size` to the certain value (see [here](https://github.com/openucx/ucx/compare/master...ivankochin:ucx:topic/uct/fix-explicit-odp-initialization?expand=1#diff-716148bcb0d4d7e0ef6be8f54003f7cc322baaca473baf9a442182c42830b7f8L768))
2. `uct_ib_md_open_common()` overwrites `md->super.config` by `md_config->ext` that contain the actual config values and therefore always set `md->config.odp.max_size` to 0. (see [here](https://github.com/openucx/ucx/compare/master...ivankochin:ucx:topic/uct/fix-explicit-odp-initialization?expand=1#diff-975534026ea952036f6617e7a106314af680d3ca405d777ab04e0af2b427499aR1832))

Solution:
1. Remove odp_max_size at all
2. Set corresponding property isntead of that

### CODE REVIEW

**Comment by brminich** [comment]
pls check if the failure is relevant:
```
2023-02-14T18:01:26.1244935Z [----------] 1 test from test_rcache_pfn, where TypeParam = 
2023-02-14T18:01:26.1247850Z [ RUN      ] test_rcache_pfn.enum_pfn <> <>
2023-02-14T18:01:26.5044859Z /scrap/azure/agent-02/AZP_WORKSPACE/1/s/contrib/../test/gtest/ucs/test_rcache.cc:1004: Failure
2023-02-14T18:01:26.5048074Z Error: Input/output error
2023-02-14T18:01:26.5118706Z [    PAGES ] 1 2 4 8 16 32 64 128 256 512 1024 2048 4096 8192 16384 32768 
2023-02-14T18:01:26.5122760Z [  FAILED  ] test_rcache_pfn.enum_pfn, where TypeParam =  and GetParam() =  (388 ms)
2023-02-14T18:01:26.5127017Z [----------] 1 test from test_rcache_pfn (388 ms total)
```

File: src/uct/api/uct.h [hunk 1]
```diff
@@ -759,7 +759,12 @@ enum {
     /**
      * MD supports registering a dmabuf file descriptor.
      */
-    UCT_MD_FLAG_REG_DMABUF    = UCS_BIT(10)
+    UCT_MD_FLAG_REG_DMABUF    = UCS_BIT(10),
+
+    /**
+     * MD supports registering on demand paging memory.
+     */
+    UCT_MD_FLAG_REG_ODP       = UCS_BIT(11)
```
Comment (yosefe) [inline src/uct/api/uct.h:767]: UCT_MD_FLAG_REG_NONBLOCK

File: src/uct/ib/base/ib_md.c [hunk 2]
```diff
@@ -1835,7 +1835,8 @@ ucs_status_t uct_ib_md_open_common(uct_ib_md_t *md,
                           UCT_MD_FLAG_NEED_RKEY |
                           UCT_MD_FLAG_ADVISE;
 
-    if (md->config.odp.max_size == UCS_MEMUNITS_AUTO) {
+    if ((md->config.odp.max_size == UCS_MEMUNITS_AUTO) &&
+        !(md->cap_flags & UCT_MD_FLAG_REG_ODP)) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c:1839]: can we remove md->config.odp.max_size altogether?
it's a reminiscent from older devices.
we can say ODP does not have size limit when supported. so if umr_extended_translation_offset is 0, ODP is not supported.

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 3]
```diff
@@ -1227,6 +1228,9 @@ static ucs_status_t uct_ib_mlx5_devx_md_open(struct ibv_device *ibv_device,
         goto err_lru_cleanup;
     }
 
+    md->super.config.odp.max_size = ucs_min(md->super.config.odp.max_size,
```
Comment (yosefe) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:1231]: also remove md->super.config.odp.max_size

File: src/uct/api/uct.h (hunk 1 above)
Comment (yosefe) [reply src/uct/api/uct.h:767]: unresolved

File: src/uct/api/uct.h [hunk 4]
```diff
@@ -759,7 +759,12 @@ enum {
     /**
      * MD supports registering a dmabuf file descriptor.
      */
-    UCT_MD_FLAG_REG_DMABUF    = UCS_BIT(10)
+    UCT_MD_FLAG_REG_DMABUF    = UCS_BIT(10),
+
+    /**
+     * MD supports registering on demand paging memory.
+     */
+    UCT_MD_FLAG_REG_NONBLOCK       = UCS_BIT(11)
```
Comment (brminich) [inline src/uct/api/uct.h:767]: alignment

File: src/uct/api/uct.h [hunk 5]
```diff
@@ -759,7 +759,12 @@ enum {
     /**
      * MD supports registering a dmabuf file descriptor.
      */
-    UCT_MD_FLAG_REG_DMABUF    = UCS_BIT(10)
+    UCT_MD_FLAG_REG_DMABUF    = UCS_BIT(10),
+
+    /**
+     * MD supports registering on demand paging memory.
```
Comment (brminich) [inline src/uct/api/uct.h:765]: MD supports non-blocking memory registration

Comment (ivankochin) [reply src/uct/api/uct.h:765]:
```suggestion
     * MD supports non-blocking memory registration.
```

File: src/uct/api/uct.h (hunk 4 above)
Comment (ivankochin) [reply src/uct/api/uct.h:767]:
```suggestion
    UCT_MD_FLAG_REG_NONBLOCK  = UCS_BIT(11)
```

File: src/uct/api/uct.h [hunk 6]
```diff
@@ -759,7 +759,12 @@ enum {
     /**
      * MD supports registering a dmabuf file descriptor.
      */
-    UCT_MD_FLAG_REG_DMABUF    = UCS_BIT(10)
+    UCT_MD_FLAG_REG_DMABUF    = UCS_BIT(10),
+
+    /**
+     * MD supports non-blocking memory registration.
+     */
+    UCT_MD_FLAG_REG_NONBLOCK  = UCS_BIT(11)
```
Comment (brminich) [inline src/uct/api/uct.h:767]: if we arelly need this new flag it should go to the PR with UCT API

File: src/uct/sm/scopy/knem/knem_md.c [hunk 7]
```diff
@@ -203,6 +203,7 @@ ucs_status_t uct_knem_md_query(uct_md_h uct_md, uct_md_attr_v2_t *md_attr)
         md_attr->flags         |= UCT_MD_FLAG_REG;
         md_attr->reg_mem_types |= UCS_BIT(UCS_MEMORY_TYPE_HOST);
     }
+    md_attr->reg_nonblock_mem_types = 0;
```
Comment (brminich) [inline src/uct/sm/scopy/knem/knem_md.c:206]: why is this needed?

File: test/gtest/uct/ib/test_ib_md.cc [hunk 8]
```diff
@@ -171,6 +171,41 @@ UCS_TEST_P(test_ib_md, ib_md_umr_direct, "REG_METHODS=direct") {
     ib_md_umr_check(&rkey_buffer[0], false);
 }
 
+UCS_TEST_SKIP_COND_P(test_ib_md, explicit_odp_reg_advise,
+                     !check_caps(UCT_MD_FLAG_REG | UCT_MD_FLAG_ADVISE),
+                     "MLX5_DEVX_OBJECTS=dct,dcsrq")
+{
+    test_reg_advise(UCS_KBYTE, UCS_KBYTE);
+}
```
Comment (brminich) [inline test/gtest/uct/ib/test_ib_md.cc:179]: what does it check?

File: src/uct/sm/scopy/knem/knem_md.c (hunk 7 above)
Comment (ivankochin) [reply src/uct/sm/scopy/knem/knem_md.c:206]: I just discovered that during this query function we do initialize `reg_mem_types` but do not initialize `reg_nonblock_mem_types` so I decided to add the initialization.

File: test/gtest/uct/ib/test_ib_md.cc (hunk 8 above)
Comment (ivankochin) [reply test/gtest/uct/ib/test_ib_md.cc:179]: During this test we just test memory registration + advise for the ODP. I do not know do we have guarantee that after advise for the ODP memory region pinned memory won't increase so there is no check.

But maybe it would be better to do memory registration, then check the pinned memory and then call the advise verb. WDYT?

File: test/gtest/uct/ib/test_ib_md.cc [hunk 9]
```diff
@@ -171,6 +171,41 @@ UCS_TEST_P(test_ib_md, ib_md_umr_direct, "REG_METHODS=direct") {
...
+UCS_TEST_SKIP_COND_P(test_ib_md, explicit_odp_reg_mem_usage,
+                     !check_caps(UCT_MD_FLAG_REG | UCT_MD_FLAG_ADVISE),
+                     "MLX5_DEVX_OBJECTS=dct,dcsrq")
+{
+    constexpr auto size = UCS_KBYTE;
+    ucs_status_t status;
+    uct_mem_h memh;
+
+    if (!(md_attr().reg_nonblock_mem_types & UCS_BIT(UCS_MEMORY_TYPE_HOST))) {
+        UCS_TEST_SKIP_R("HW does not support atomic indirect MR");
```
[the remark this answers is not in the document]
Comment (ivankochin) [reply test/gtest/uct/ib/test_ib_md.cc:190]: My bad, thought that I fixed it.

```suggestion
        UCS_TEST_SKIP_R("HW does not support non-blocking memory registration");
```

File: test/gtest/uct/ib/test_ib_md.cc (hunk 8 above)
Comment (ivankochin) [reply test/gtest/uct/ib/test_ib_md.cc]: I overthought that and decided to move all the tests to the `test_md.cc` file since it seems more about all the MDs now, not only about the IB. Pls review.

File: test/gtest/uct/test_md.cc [hunk 10]
```diff
@@ -621,15 +621,21 @@ UCS_TEST_SKIP_COND_P(test_md, reg_perf,
...
+{
+    int vmpin_before, vmpin_after;
     ucs_status_t status;
     void *address;
     uct_mem_h memh;
 
-    size = 128 * UCS_MBYTE;
+    if (check_non_blocking) {
+        if (!(md_attr().reg_nonblock_mem_types & UCS_BIT(UCS_MEMORY_TYPE_HOST))) {
+            UCS_TEST_SKIP_R("HW does not support non-blocking registration");
```
Comment (brminich) [inline test/gtest/uct/test_md.cc:634]: minor: `s/HW/MD`

File: test/gtest/uct/test_md.cc [hunk 11]
```diff
@@ -638,16 +644,28 @@ UCS_TEST_SKIP_COND_P(test_md, reg_advise,
...
+                                   advise_size, UCT_MADV_WILLNEED);
+        EXPECT_UCS_OK(status);
+    }
 
     status = uct_md_mem_dereg(md(), memh);
     EXPECT_UCS_OK(status);
     free(address);
 }
 
+UCS_TEST_SKIP_COND_P(test_md, reg_advise,
```
Comment (brminich) [inline test/gtest/uct/test_md.cc:663]: why need this test now? It can be a part of test_md_non_blocking

File: test/gtest/uct/test_md.cc [hunk 12]
```diff
@@ -989,3 +1007,29 @@ UCS_TEST_P(test_md_memlock_limit, md_open)
...
+        test_md::init();
+    }
+};
+
+UCS_TEST_SKIP_COND_P(test_md_non_blocking, reg_advise,
+                     !check_caps(UCT_MD_FLAG_REG | UCT_MD_FLAG_ADVISE))
+{
+    test_reg_advise(UCS_KBYTE, UCS_KBYTE, 0, true);
+    test_reg_advise(UCS_KBYTE, UCS_KBYTE / 2, 0, true);
+    test_reg_advise(UCS_KBYTE, UCS_KBYTE / 2, UCS_KBYTE / 4, true);
```
Comment (brminich) [inline test/gtest/uct/test_md.cc:1026]: why are all the sizes that small? Imo, better to use multiple sizes, so that we could test many pages

Comment (ivankochin) [reply test/gtest/uct/test_md.cc:1026]: Due to internal bug that broke ODPv1 in case of huge size and allocation through malloc (not memalign).

File: test/gtest/uct/test_md.cc (hunk 11 above)
Comment (ivankochin) [reply test/gtest/uct/test_md.cc:663]: I just thought that it can make sense to leave it as is and test nonblock flag with the default configuration.

File: test/gtest/uct/test_md.cc (hunk 10 above)
Comment (ivankochin) [reply test/gtest/uct/test_md.cc:634]:
```suggestion
            UCS_TEST_SKIP_R("MD does not support non-blocking registration");
```

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 13]
```diff
@@ -761,20 +761,8 @@ uct_ib_mlx5_devx_check_odp(uct_ib_mlx5_md_t *md,
...
-        } else {
-            md->super.config.odp.max_size = 1ul << 28;
-        }
-
-        md->super.reg_nonblock_mem_types = UCS_BIT(UCS_MEMORY_TYPE_HOST);
-    }
-
-    return UCS_OK;
-
+    md->super.reg_nonblock_mem_types = UCS_BIT(UCS_MEMORY_TYPE_HOST);
```
Comment (yosefe) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:764]: if umr_extended_translation_offset is 0, ODP should be considered unsupported

File: test/gtest/common/test_helpers.cc [hunk 14]
```diff
@@ -469,6 +469,26 @@ bool is_interface_usable(struct ifaddrs *ifa)
...
+
+int get_proc_self_status_field(const std::string &parameter)
+{
+    const std::string path("/proc/self/status");
+    std::ifstream proc_stats(path);
+    std::string line, name;
+    int value;
+
+    while (std::getline(proc_stats, line)) {
+        if (!(std::istringstream(line) >> name >> value)) {
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc:481]: extra ( )

File: test/gtest/common/test_helpers.cc [hunk 15]
```diff
@@ -469,6 +469,26 @@ bool is_interface_usable(struct ifaddrs *ifa)
...
+            continue;
+        }
+        if (name.find(parameter) != std::string::npos) {
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc:484]: maybe use == ?

File: test/gtest/common/test_helpers.cc [hunk 16]
```diff
@@ -469,6 +469,26 @@ bool is_interface_usable(struct ifaddrs *ifa)
...
+            return value;
+        }
+    }
+    UCS_TEST_MESSAGE << path << " does not contain " << parameter << " value";
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc:488]: space line before

File: test/gtest/common/test_helpers.cc [hunk 17]
```diff
@@ -469,6 +469,26 @@ bool is_interface_usable(struct ifaddrs *ifa)
            !netif_has_sysfs_file(ifa->ifa_name, "wireless");
 }
 
+
+int get_proc_self_status_field(const std::string &parameter)
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc:473]: size_t / ssize_t

File: test/gtest/common/test_helpers.cc [hunk 18]
```diff
@@ -469,6 +469,26 @@ bool is_interface_usable(struct ifaddrs *ifa)
...
+    while (std::getline(proc_stats, line)) {
+        if (!(std::istringstream(line) >> name >> value)) {
+            continue;
+        }
+        if (name.find(parameter) != std::string::npos) {
+            return value;
+        }
+    }
+    UCS_TEST_MESSAGE << path << " does not contain " << parameter << " value";
+    return -1;
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc:489]: maybe 0 is a better default value?
or pass default value as input param to this function?

File: test/gtest/uct/test_md.cc [hunk 19]
```diff
@@ -621,15 +621,21 @@ UCS_TEST_SKIP_COND_P(test_md, reg_perf,
...
     ucs_status_t status;
     void *address;
     uct_mem_h memh;
 
-    size = 128 * UCS_MBYTE;
+    if (check_non_blocking) {
+        if (!(md_attr().reg_nonblock_mem_types & UCS_BIT(UCS_MEMORY_TYPE_HOST))) {
+            UCS_TEST_SKIP_R("MD does not support non-blocking registration");
+        }
+        vmpin_before = ucs::get_proc_self_status_field("VmPin");
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc:636]: space line before

File: test/gtest/uct/test_md.cc [hunk 20]
```diff
@@ -638,16 +644,28 @@ UCS_TEST_SKIP_COND_P(test_md, reg_advise,
...
+    if (check_non_blocking) {
+        vmpin_after = ucs::get_proc_self_status_field("VmPin");
+        ASSERT_EQ(vmpin_before, vmpin_after);
+    }
 
-    status = uct_md_mem_advise(md(), memh, (char *)address + 7,
-                               32 * UCS_KBYTE, UCT_MADV_WILLNEED);
-    EXPECT_UCS_OK(status);
+    if (advise_size) {
+        status = uct_md_mem_advise(md(), memh, (char*)address + advice_offset,
```
Comment (yosefe) [inline test/gtest/uct/test_md.cc:653]: use UCS_PTR_BYTE_OFFSET

File: test/gtest/uct/test_md.cc (hunk 12 above)
Comment (yosefe) [reply test/gtest/uct/test_md.cc:1026]: Which size was it?

Comment (ivankochin) [reply test/gtest/uct/test_md.cc:1026]: 32K+

All details are here: https://redmine.mellanox.com/issues/3363764 (internal link)

File: test/gtest/common/test_helpers.cc (hunk 14 above)
Comment (ivankochin) [reply test/gtest/common/test_helpers.cc:481]: `!` has higher precedence than `>>` (https://en.cppreference.com/w/cpp/language/operator_precedence)

File: test/gtest/common/test_helpers.cc (hunk 15 above)
Comment (ivankochin) [reply test/gtest/common/test_helpers.cc:484]: Do you mean something like that?

```c++
if (name. find(parameter) == std::string::npos) {
    continue;
}
return value;
```

File: test/gtest/common/test_helpers.cc (hunk 18 above)
Comment (ivankochin) [reply test/gtest/common/test_helpers.cc:489]: As I understand `0` is the valid value for this function and I use `-1` as a signal that the value wasn't found at all. I can return `0` and print some message in that case if you prefer so.

File: test/gtest/uct/test_md.cc [hunk 21]
```diff
@@ -989,3 +1007,29 @@ UCS_TEST_P(test_md_memlock_limit, md_open)
 }
 
 UCT_MD_INSTANTIATE_TEST_CASE(test_md_memlock_limit)
+
+class test_md_non_blocking : public test_md
+{
+protected:
+    void init() override {
+        /* ODPv1 IB feature can work only for certain DEVX configuration */
```
[the remark this answers is not in the document]
Comment (ivankochin) [reply test/gtest/uct/test_md.cc:1017]: Yes, it will be 0

File: test/gtest/uct/test_md.cc (hunk 12 above)
Comment (brminich) [reply test/gtest/uct/test_md.cc:1028]: can you please add the test with bigger message size, but comment it for now? Also can mention that it needs to be uncomented once this RM is fixed.

File: test/gtest/common/test_helpers.cc (hunk 15 above)
Comment (yosefe) [reply test/gtest/common/test_helpers.cc:484]: if (name == (parameter + ":"))

File: test/gtest/common/test_helpers.cc (hunk 17 above)
Comment (ivankochin) [reply test/gtest/common/test_helpers.cc:473]: I use `int` to return `-1` error code.

File: test/gtest/uct/test_md.cc [hunk 22]
```diff
@@ -621,15 +621,23 @@ UCS_TEST_SKIP_COND_P(test_md, reg_perf,
...
     uct_mem_h memh;
 
-    size = 128 * UCS_MBYTE;
+    if (check_non_blocking) {
+        if (!(md_attr().reg_nonblock_mem_types & UCS_BIT(UCS_MEMORY_TYPE_HOST))) {
+            UCS_TEST_SKIP_R("MD does not support non-blocking registration");
+        }
+
+        vmpin_before = ucs::get_proc_self_status_field("VmPin");
+        ASSERT_TRUE(vmpin_before != -1);
```
Comment (brminich) [inline test/gtest/uct/test_md.cc:638]: minor: can use `ASSERT_NE`

**Review by brminich — APPROVED** [review]

File: test/gtest/uct/test_md.cc (hunk 22 above)
Comment (ivankochin) [reply test/gtest/uct/test_md.cc:638]:
```suggestion
        ASSERT_NE(vmpin_before, -1);
```

File: test/gtest/common/test_helpers.cc (hunk 17 above)
Comment (brminich) [reply test/gtest/common/test_helpers.cc:473]: btw can use `ssize_t` as Yossi commented

**Review by brminich — DISMISSED** [review]

**Review by yosefe — DISMISSED** [review]

**Comment by yosefe** [comment]
@ivankochin pls squash

**Comment by ivankochin** [comment]
> @ivankochin pls squash

done

**Review by brminich — APPROVED** [review]
